closed/make/Javadoc.gmk

```diff
@@ -0,0 +1,490 @@
+# following files:
+# 1. The openj9 java source files which override the openjdk versions
+# 2. The openj9.xxxx modules, including the javadoc generated files such as
+#    enum class pages, module-summary.html, package-summary.html etc.)
+# A temporary script file is used to avoid the xargs command line length limit
+# of 255 chars.
+
+OPENJ9_SUBST_BOTTOM = $(subst &,\&,$(OPENJ9_JAVADOC_BOTTOM))
+
+#############################################################################
+# sed command to replace the footer in the jdk management javadoc files.
+MGMT_BOTTOM_REGEX = $(subst .,\.,$(MGMT_BOTTOM))
+OPENJ9_MGMT_SUBST_COMMAND = $(SED) -e '\''s|$(MGMT_BOTTOM_REGEX)|$(OPENJ9_SUBST_BOTTOM)|g'\''
+
+#############################################################################
+# sed command to replace the footer in the core api javadoc files.
+# Openjdk Core API doc is generated with the -Xdocrootparent javadoc option,
+# so apply the same substition to the page footer before matching.
+COREAPI_BOTTOM_REGEX = $(subst {@docroot}/,$(subst .,\.,$(DOCS_BASE_URL)),$(subst .,\.,$(COREAPI_BOTTOM)))
+OPENJ9_COREAPI_SUBST_COMMAND = $(SED) -e '\''s|$(COREAPI_BOTTOM_REGEX)|$(OPENJ9_SUBST_BOTTOM)|g'\''
+
+################################################################################
+OPENJ9_JAVADOC_MARKER := $(TEMPDIR)/_javadoc_openj9.marker
+OPENJ9_JAVADOC_SCRIPT := $(TEMPDIR)/_javadoc_openj9_footer_replace.sh
+OPENJ9_JAVADOC_TEMP  := $(TEMPDIR)/_javadoc_temp.html
+
+################################################################################
+# Use a marker file as target with dependencies which ensure footers are not
+# updated until the pages have been updated by the openjdk javadoc targets. 
+
+OPENJ9_JAVADOC_SCRIPT_BODY = \
+    for path in "$$@" ; do \
+      if [ -f "$(COREAPI_DOCSDIR)/$$path" ] ; then \
+        $(CP) "$(COREAPI_DOCSDIR)/$$path" "$(OPENJ9_JAVADOC_TEMP)" ; \
+        $(OPENJ9_COREAPI_SUBST_COMMAND) < "$(OPENJ9_JAVADOC_TEMP)" > "$(DOCS_OUTPUTDIR)/$$path" ; \
+        $(RM) "$(OPENJ9_JAVADOC_TEMP)" ; \
+      fi ; \
+      if [ -f "$(JRE_API_DOCSDIR)/management/extension/$$path" ] ; then \
+        $(CP) "$(JRE_API_DOCSDIR)/management/extension/$$path" "$(OPENJ9_JAVADOC_TEMP)" ; \
+        $(OPENJ9_MGMT_SUBST_COMMAND) < "$(OPENJ9_JAVADOC_TEMP)" > "$(JRE_API_DOCSDIR)/management/extension/$$path" ; \
+        $(RM) "$(OPENJ9_JAVADOC_TEMP)" ; \
+      fi ; \
+    done \
+    #
+
+$(OPENJ9_JAVADOC_MARKER) : $(COREAPI_DOCSDIR)/index.html openj9-mgmtdocs
+	$(ECHO) '$(OPENJ9_JAVADOC_SCRIPT_BODY)' >$(OPENJ9_JAVADOC_SCRIPT)
+	$(CD) $(SUPPORT_OUTPUTDIR)/j9jcl_sources ; \
+		$(FIND) . -type f -name '*.java' \
+			| $(SED) -e 's|/[^/]\+/classes/|/|' -e 's|\.java$$|.html|' \
+			| $(XARGS) $(BASH) $(OPENJ9_JAVADOC_SCRIPT)
+	$(CD) $(TOPDIR)/closed/src ; \
+		$(FIND) . -type f -name '*.java' \
+			| $(SED) -e 's|/[^/]\+/classes/|/|' -e 's|\.java$$|.html|' \
+			| $(XARGS) $(BASH) $(OPENJ9_JAVADOC_SCRIPT)
+	$(CD) $(COREAPI_DOCSDIR) ; \
```

Comment:
Do any of `JDK_API_DOCSDIR`, `JRE_API_DOCSDIR` or `PLATFORM_DOCSDIR` need to be processed?

Reply:
The code under review was incorrect and should now be fixed.
But to answer the question:
JDK_API_DOCSDIR - contains javadoc for attach, javac, javadoc, jconsole, jpda, nashorn.  I do not believe these APIs are modified by openj9
JRE_API_DOCSDIR - contains javadoc for accessibility, management, net, nio, plugin, security.  We need to process JRE_API_DOCSDIR/management/extension because openj9 provides its own management classes
PLATFORM_DOCSDIR - contains javadoc for the openj9 specific packages, but we don't need to post process these files because the footer is set correctly on the javadoc commands (the issue with some files needing openjdk footers and some needing openj9 footers doesn't arise).

Reply:
I believe there are openj9 contributions to all of those groups:
* jcmd, jps should appear in JDK_API_DOCSDIR
* JRE_API_DOCSDIR should include java.base classes - see the next point
* if any java.base classes appear in JDK_API_DOCSDIR, they may need fixing (e.g. java.lang.String)

Reply:
1. javadoc is not produced for jcmd and jps by the openjdk build - that is why we no not see the javadoc for the openj9 jdk.  
2. The base class modifications are included in the coredocs javadoc (docs/api).  Only the openj9.xx packages are included under PLATFORM_DOCSDIR.

Reply:
I didn't do a thorough investigation, but noticed no OpenJ9 footers anywhere other than in JDK_API_DOCSDIR. That's fine if there shouldn't be anything there, but it was surprising, and I would expect to se jcmd and jps, for example, documented somewhere.

I guess the summary, from my point of view, is that I don't see the structure in the makefiles to deal with future changes to Java code in OpenJ9 that should land in those extra folders. Please help me see it or add it.

Reply:
The openj9 overrides to the core packages replace the openjdk ones under docs/api - so for instance docs/api/java/lang/String.html contains the javadoc and the footer from the openj9 String class override.
The sources files are identified by the find command:
```
	$(CD) $(JDK_OUTPUTDIR)/j9jcl_sources/jdk/src ; \
		$(FIND) . -type f -name '*.java' \
			| $(SED) -e 's|/[^/]\+/classes/|/|' -e 's|\.java$$|.html|' \
			| $(XARGS) $(BASH) $(OPENJ9_JAVADOC_SCRIPT)
```
and the corresponding html file footers replaced by the
```
if [ -f "$(COREAPI_DOCSDIR)/$$path" ] ; then \
        $(CP) "$(COREAPI_DOCSDIR)/$$path" "$(OPENJ9_JAVADOC_TEMP)" ; \
        $(OPENJ9_COREAPI_SUBST_COMMAND) < "$(OPENJ9_JAVADOC_TEMP)" > "$(COREAPI_DOCSDIR)/$$path" ; \
        $(RM) "$(OPENJ9_JAVADOC_TEMP)" ; \
      fi ;
```
statement in the temporary script file.